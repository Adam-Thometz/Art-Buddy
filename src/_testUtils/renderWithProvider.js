import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

// import renderWithRouter from './renderWithRouter';
import rootReducer from '../_redux/reducers/rootReducer';

function withProvider({ children, store }) {
  return (
    <Provider store={store}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
};

function renderWithProvider(component, options) {
  const store = (options && options.store)
    ? options.store
    : rootReducer;
  return render(component, { wrapper: (args) =>
    withProvider({
      ...args,
      store
    }), ...options
  });
};

export default renderWithProvider;