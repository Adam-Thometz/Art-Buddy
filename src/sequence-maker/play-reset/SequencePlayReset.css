.SequencePlayReset {
  display: flex;
  justify-content: center;
}

.SequencePlayReset-play {
  display: flex;
  width: 15%;
  justify-content: space-evenly;
}

.SequencePlayReset-reset {
  display: flex;
  width: 20%;
  justify-content: center;
}