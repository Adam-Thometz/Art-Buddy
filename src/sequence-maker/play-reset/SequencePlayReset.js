import React from 'react';

import { useSelector, useDispatch } from "react-redux";
import { resetSequence, togglePlaying } from "_redux/sequence-maker/sequenceMakerActions";

import './SequencePlayReset.css';

import Icon from "_components/icon/Icon";

import { playSequence } from "_utils/sequence-maker/playSequence";
import play from '_media/sequence-maker/_icons/play.png';
import playAll from '_media/sequence-maker/_icons/play-all.png';
import reset from '_media/sequence-maker/_icons/reset.png';
import { start, Transport } from 'tone';

const SequencePlayReset = () => {
  const {
    sequence,
    pitch,
    duration
  } = useSelector(state => state.sequenceMaker);
  const { volume } = useSelector(state => state.mainSettings);
  const dispatch = useDispatch(); 

  const handlePlay = async () => {
    if (Transport.state === 'stopped') await start();
    const isPlaying = sequence
      .filter(block => block !== null)
      .some(block => block.isPlaying)
    if (isPlaying) return;

    for (let i = 0; i < sequence.length; i++) {
      if (!sequence[i]) continue;
      const start = setTimeout(() => {
        dispatch(togglePlaying(i));
        clearTimeout(start);
      }, (duration * 1000) * i);
      const end = setTimeout(() => {
        dispatch(togglePlaying(i));
        clearTimeout(end);
      }, (duration * 1000) + ((duration * 1000) * i));
    };
    playSequence({ sequence, pitch, duration, volume, playAll: false });
  };
  
  const handlePlayAll = async () => {
    if (Transport.state === 'stopped') await start();
    const isPlaying = sequence
      .filter(block => block !== null)
      .some(block => block.isPlaying)
    if (isPlaying) return;

    for (let i = 0; i < sequence.length; i++) {
      if (!sequence[i]) continue;
      dispatch(togglePlaying(i));
      const timer = setTimeout(() => {
        dispatch(togglePlaying(i));
        clearTimeout(timer);
      }, duration * 1000);
    };
    playSequence({ sequence, pitch, duration, volume, playAll: true });
  };

  const handleReset = () => {
    dispatch(resetSequence());
    Transport.stop();
  };
  
  return (
    <section className='SequencePlayReset'>
      <aside className="SequencePlayReset-play">
        <Icon icon={play} text='Play' size='48px' onClick={handlePlay} />
        <Icon icon={playAll} text='Play All' size='48px' onClick={handlePlayAll} />
      </aside>
      <aside className="SequencePlayReset-reset">
        <Icon icon={reset} text='Reset Sequence' size="48px" onClick={handleReset} />
      </aside>
    </section>
  );
};

export default SequencePlayReset;