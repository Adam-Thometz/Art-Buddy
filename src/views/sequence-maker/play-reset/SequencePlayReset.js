import { useContext } from "react";
import { PlayContext } from "_context/PlayContext";

import { useSelector, useDispatch } from "react-redux";
import { resetSequence, togglePlaying } from "_redux/sequence-maker/sequenceMakerReducer";

import "./SequencePlayReset.css";

import Icon from "_components/icon/Icon";

import { PLAY, PLAY_ALL, RESET } from "_assets/sequence-maker/sequenceMaker.assets";
import { start, Transport } from "tone";

const SequencePlayReset = () => {
  const { sequence, pitch, duration, isPlaying } = useSelector(
    (state) => state.sequenceMaker
  );
  const { playFn } = useContext(PlayContext);
  const dispatch = useDispatch();

  const handlePlay = async () => {
    if (Transport.state === "stopped") await start();
    if (isPlaying) return;

    for (let i = 0; i < sequence.length; i++) {
      if (!sequence[i]) continue;
      const start = setTimeout(() => {
        dispatch(togglePlaying(i));
        clearTimeout(start);
      }, duration * 1000 * i);
      const end = setTimeout(() => {
        dispatch(togglePlaying(i));
        clearTimeout(end);
      }, duration * 1000 + duration * 1000 * i);
    }
    playFn.playSequence({ pitch, duration, playAll: false });
  };

  const handlePlayAll = async () => {
    if (Transport.state === "stopped") await start();
    if (isPlaying) return;

    for (let i = 0; i < sequence.length; i++) {
      if (!sequence[i]) continue;
      dispatch(togglePlaying(i));
      const timer = setTimeout(() => {
        dispatch(togglePlaying(i));
        clearTimeout(timer);
      }, duration * 1000);
    }
    playFn.playSequence({ pitch, duration, playAll: true });
  };

  const handleReset = () => {
    dispatch(resetSequence());
    Transport.stop();
  };

  return (
    <section className="SequencePlayReset">
      <aside className="SequencePlayReset-play">
        <Icon
          icon={PLAY}
          text="Play"
          size="48px"
          onClick={handlePlay}
          testId="play"
        />
        <Icon
          icon={PLAY_ALL}
          text="Play All"
          size="48px"
          onClick={handlePlayAll}
          testId="playAll"
        />
      </aside>
      <aside className="SequencePlayReset-reset">
        <Icon
          icon={RESET}
          text="Reset Sequence"
          size="48px"
          onClick={handleReset}
          testId="reset"
        />
      </aside>
    </section>
  );
};

export default SequencePlayReset;
