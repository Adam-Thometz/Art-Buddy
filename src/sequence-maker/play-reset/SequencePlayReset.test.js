import React from "react";
import renderWithProvider from "_testUtils/renderWithProvider";
import SequencePlayReset from './SequencePlayReset';

describe('SequencePlayReset component', () => {
  it('renders without crashing', () => {
    renderWithProvider(<SequencePlayReset />);
  });

  it('matches the snapshot of SequencePlayReset', () => {
    const { asFragment } = renderWithProvider(<SequencePlayReset />);
    expect(asFragment()).toMatchSnapshot();
  });
});