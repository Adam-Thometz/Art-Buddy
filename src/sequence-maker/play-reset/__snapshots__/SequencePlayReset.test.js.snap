// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequencePlayReset component matches the snapshot of SequencePlayReset 1`] = `
<DocumentFragment>
  <section
    class="SequencePlayReset"
  >
    <aside
      class="SequencePlayReset-play"
    >
      <figure
        class="Icon"
        data-testid=""
      >
        <img
          alt=""
          src="file"
          style="height: 48px; width: 48px; cursor: pointer;"
        />
        <figcaption
          class="Icon-text"
          style="font-size: 1rem;"
        >
          Play
        </figcaption>
      </figure>
      <figure
        class="Icon"
        data-testid=""
      >
        <img
          alt=""
          src="file"
          style="height: 48px; width: 48px; cursor: pointer;"
        />
        <figcaption
          class="Icon-text"
          style="font-size: 1rem;"
        >
          Play All
        </figcaption>
      </figure>
    </aside>
    <aside
      class="SequencePlayReset-reset"
    >
      <figure
        class="Icon"
        data-testid=""
      >
        <img
          alt=""
          src="file"
          style="height: 48px; width: 48px; cursor: pointer;"
        />
        <figcaption
          class="Icon-text"
          style="font-size: 1rem;"
        >
          Reset Sequence
        </figcaption>
      </figure>
    </aside>
  </section>
</DocumentFragment>
`;
